sdks/python/apache_beam/runners/worker/bundle_processor.py

```diff
@@ -1264,14 +1275,28 @@ def process(
 def create(*args):
 
   class SplitAndSizeRestrictions(beam.DoFn):
-    def __init__(self, fn, restriction_provider):
+    def __init__(self,
+                 fn,
+                 restriction_provider,
+                 watermark_estimator_provider=None):
       self.restriction_provider = restriction_provider
+      self.watermark_estimator_provider = watermark_estimator_provider
 
     def process(self, element_restriction, *args, **kwargs):
-      element, restriction = element_restriction
-      for part, size in self.restriction_provider.split_and_size(
-          element, restriction):
-        yield ((element, part), size)
+      if self.watermark_estimator_provider:
+        element, (restriction, _) = element_restriction
+        for part, size in self.restriction_provider.split_and_size(
+            element, restriction):
+          yield ((element,
+                  (part,
+                   self.watermark_estimator_provider.initial_estimator_state(
```

Comment:
Note that I call `initial_estimator_state(element, splitted_restriction)` again given that restriction could be changed by initial splitting.